midend/flattenLogMsg.h

```diff
@@ -0,0 +1,92 @@
+#ifndef _MIDEND_FLATTENILOGMSG_H_
+#define _MIDEND_FLATTENILOGMSG_H_
+
+#include "ir/ir.h"
+#include "frontends/p4/typeChecking/typeChecker.h"
+
+namespace P4 {
+
+using TypeLogMsgParams = std::pair<IR::IndexedVector<IR::NamedExpression>, std::string>;
+
+/*
+   Find types in log_msg function for replace.
+*/
+class FindTypeInLogMsgForReplace : public Inspector {
+   P4::TypeMap* typeMap;
+   ordered_map<cstring, const IR::Type_StructLike*> replacement;
+   ordered_map<unsigned, const IR::MethodCallStatement* > logMsgReplacament;
+   size_t index;
+
+ public:
+    explicit FindTypeInLogMsgForReplace(P4::TypeMap *typeMap) : typeMap(typeMap) {
+        setName("FindTypeInLogMsgForReplace");
+        CHECK_NULL(typeMap);
+    }
+    bool preorder(const IR::MethodCallStatement* methodCallStatement) override;
+    void createReplacement(const IR::Type_StructLike* type);
+    const IR::MethodCallStatement* prepareLogMsgStatement(
+         const IR::MethodCallStatement* methodCallStatement);
+    const IR::Type_StructLike* getReplacement(const cstring name) const {
+         return ::get(replacement, name); }
+    const IR::MethodCallStatement* getReplacementMethodCall(unsigned id) const {
+       return ::get(logMsgReplacament, id);
+    }
+    bool empty() const { return replacement.empty(); }
+   const IR::Type_StructLike* hasStructInParameter(
```

Comment:
it seems to be enough to return a boolean here, you don't use the type anywhere.